url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/domain/singleagent/sokoban2/Sokoban2Visualizer.java

```diff
@@ -0,0 +1,483 @@
+			this.maxY = maxY;
+		}
+		
+		@Override
+		public void paintObject(Graphics2D g2, State s, ObjectInstance ob, float cWidth, float cHeight) {
+			
+			float domainXScale = Sokoban2Domain.maxRoomXExtent(s) + 1f;
+			float domainYScale = Sokoban2Domain.maxRoomYExtent(s) + 1f;
+			
+			if(maxX != -1){
+				domainXScale = maxX;
+				domainYScale = maxY;
+			}
+			
+			//determine then normalized width
+			float width = (1.0f / domainXScale) * cWidth;
+			float height = (1.0f / domainYScale) * cHeight;
+			
+			int top = ob.getDiscValForAttribute(Sokoban2Domain.ATTTOP);
+			int left = ob.getDiscValForAttribute(Sokoban2Domain.ATTLEFT);
+			int bottom = ob.getDiscValForAttribute(Sokoban2Domain.ATTBOTTOM);
+			int right = ob.getDiscValForAttribute(Sokoban2Domain.ATTRIGHT);
+			
+			g2.setColor(Color.white);
+			
+			for(int i = left; i <= right; i++){
+				for(int j = bottom; j <= top; j++){
+					
+					float rx = i*width;
+					float ry = cHeight - height - j*height;
+					g2.fill(new Rectangle2D.Float(rx, ry, width, height));
+					
+				}
+			}
+			
+			
+		}
+
+	}
+	
+	
+	
+	public static class AgentPainter implements ObjectPainter{
+
+		protected int maxX = -1;
+		protected int maxY = -1;
+		
+		public AgentPainter(){
+			
+		}
+		
+		public AgentPainter(int maxX, int maxY){
+			this.maxX = maxX;
+			this.maxY = maxY;
+		}
+		
+		@Override
+		public void paintObject(Graphics2D g2, State s, ObjectInstance ob, float cWidth, float cHeight) {
+			
+			g2.setColor(Color.darkGray);
+			
+			float domainXScale = Sokoban2Domain.maxRoomXExtent(s) + 1f;
+			float domainYScale = Sokoban2Domain.maxRoomYExtent(s) + 1f;
+			
+			if(maxX != -1){
+				domainXScale = maxX;
+				domainYScale = maxY;
+			}
+			
+			//determine then normalized width
+			float width = (1.0f / domainXScale) * cWidth;
+			float height = (1.0f / domainYScale) * cHeight;
+			
+			int x = ob.getDiscValForAttribute(Sokoban2Domain.ATTX);
+			int y = ob.getDiscValForAttribute(Sokoban2Domain.ATTY);
+			
+			float rx = x*width;
+			float ry = cHeight - height - y*height;
+			
+			g2.fill(new Rectangle2D.Float(rx, ry, width, height));
+			
+		}
+		
+		
+		
+	}
+	
+	
+	public static class AgentPainterWithImages implements ObjectPainter, ImageObserver{
+
+		protected int maxX = -1;
+		protected int maxY = -1;
+		
+		Map<String, BufferedImage>	dirToImage;
+		
+		public AgentPainterWithImages(String pathToImageDir){
+			if(!pathToImageDir.endsWith("/")){
+				pathToImageDir = pathToImageDir + "/";
+			}
+			
+			dirToImage = new HashMap<String, BufferedImage>(4);
+			try {
+				dirToImage.put("north", ImageIO.read(new File(pathToImageDir + "robotNorth.png")));
+				dirToImage.put("south", ImageIO.read(new File(pathToImageDir + "robotSouth.png")));
+				dirToImage.put("east", ImageIO.read(new File(pathToImageDir + "robotEast.png")));
+				dirToImage.put("west", ImageIO.read(new File(pathToImageDir + "robotWest.png")));
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 283:290, 302:309.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.